test/e2e/gateway_api_test.go

```diff
@@ -0,0 +1,266 @@
+//go:build e2e
+// +build e2e
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/cluster-ingress-operator/pkg/operator/controller/gatewayclass"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/storage/names"
+
+	gwapi "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	// The expected OSSM subscription name.
+	expectedSubscriptionName = "servicemeshoperator"
+	// The expected OSSM catalog source name.
+	expectedCatalogSourceName = "redhat-operators"
+	// The expected catalog source namespace.
+	expectedCatalogSourceNamespace = "openshift-marketplace"
+	// The test gateway name used in multiple places.
+	testGatewayName = "test-gateway"
+)
+
+var crdNames = []string{
+	"gatewayclasses.gateway.networking.k8s.io",
+	"gateways.gateway.networking.k8s.io",
+	"httproutes.gateway.networking.k8s.io",
+	"referencegrants.gateway.networking.k8s.io",
+}
+
+// Global variables for testing.
+// Record whether the feature gate for Gateway API is enabled.
+var gatewayAPIEnabled = false
+
+// The contents of the cluster feature gate.
+var clusterFeatureGate = &configv1.FeatureGate{}
+
+// The list of enabled features in the feature gate.
+var enabledFeatures []configv1.FeatureGateName
+
+// The default route name to be constructed.
+var default_routename = ""
+
+// If the Gateway API feature gate is enabled, run a series of tests in order
+// to validate if Gateway API resources are available, objects can be created
+// successfully and also work properly, and that the Istio installation was
+// successful.
+// NOTE: do not change the name of the test.  If new tests are added while the
+// feature gate is still in effect, preface the test names with "TestGatewayAPI"
+// so that they run via the openshift/release test configuration.
+func TestGatewayAPI(t *testing.T) {
+	getGatewayAPIEnabled(t)
+	// Skip if feature is not enabled
+	if !gatewayAPIEnabled {
+		t.Log("Gateway API not enabled, skipping TestGatewayAPI")
+		return
+	}
+
+	// Defer the cleanup of the test gateway.
+	defer func() {
+		testGateway := gwapi.Gateway{ObjectMeta: metav1.ObjectMeta{Name: testGatewayName, Namespace: openshiftIngressNamespace}}
+		if err := kclient.Delete(context.TODO(), &testGateway); err != nil {
+			if errors.IsNotFound(err) {
+				return
+			}
+			t.Errorf("failed to delete gateway %q: %v", testGateway.Name, err)
+		}
+	}()
+
+	t.Run("testGatewayAPIResources", testGatewayAPIResources)
+	t.Run("testGatewayAPIObjects", testGatewayAPIObjects)
+	t.Run("testGatewayAPIIstioInstallation", testGatewayAPIIstioInstallation)
```

Comment:
It takes a long time for all of the installation steps to come to an acceptable status, so I take the wait time waiting for the `Accepted` status of each API object, rather than waiting for each of the installation components.  Then the installation components can continue in the background while I test the next API object.